File: python/docsearch/adapters/markdown.py
"""Markdown adapter: structure from ATX heading levels."""

from __future__ import annotations

import re
from pathlib import Path

from ..blocks import Block, Extraction

_ATX = re.compile(r"^(#{1,6})\s+(.*?)\s*#*$")
_FENCE = re.compile(r"^\s*(```|~~~)")


def extract(path: Path, progress: object = None) -> Extraction:
    text = path.read_text(encoding="utf-8", errors="replace")
    stack: list[str] = []
    blocks: list[Block] = []
    para: list[str] = []
    para_offset = 0
    offset = 0
    in_fence = False

    def flush() -> None:
        nonlocal para
        body = "\n".join(para).strip()
        if body:
            blocks.append(
                Block(heading_path=list(stack), locator={"offset": para_offset}, text=body)
            )
        para = []

    for line in text.splitlines(keepends=True):
        stripped = line.rstrip("\n")
        if _FENCE.match(stripped):
            in_fence = not in_fence
            para.append(stripped)
            offset += len(line)
            continue
        m = None if in_fence else _ATX.match(stripped)
        if m:
            flush()
            level = len(m.group(1))
            del stack[level - 1 :]
            while len(stack) < level - 1:
                stack.append("")
            stack.append(m.group(2).strip())
        elif not stripped.strip():
            flush()
            para_offset = offset + len(line)
        else:
            if not para:
                para_offset = offset
            para.append(stripped)
        offset += len(line)
    flush()

    first_h1 = next((b.heading_path[0] for b in blocks if b.heading_path), None)
    return Extraction(
        title=first_h1 or path.stem,
        format="markdown",
        blocks=blocks,
        diagnostics={"structure_source": "atx_headings"},
    )

File: internal/adapter/markdown/markdown.go
// Package markdown reads a Markdown file, taking its structure from ATX
// headings. A heading inside a fenced code block is code, not structure.
// Ported from python/docsearch/adapters/markdown.py.
package markdown

import (
	"path/filepath"
	"regexp"
	"strings"
	"unicode/utf8"

	"github.com/bamsammich/docsearch/internal/domain"
	"github.com/bamsammich/docsearch/internal/pystr"
)

// Python's `\s` is Unicode whitespace and Go's is ASCII, so both patterns
// spell the class out.
var (
	atxHeading = regexp.MustCompile(
		`^(#{1,6})[` + pystr.SpaceChars + `]+(.*?)[` + pystr.SpaceChars + `]*#*$`)
	codeFence = regexp.MustCompile("^[" + pystr.SpaceChars + "]*(```|~~~)")
)

// Extract reads the Markdown file at path. Its title is the first
// top-level heading any block sits under, else the file name.
func Extract(path string) (*domain.Extraction, error) {
	raw, err := pystr.ReadFile(path)
	if err != nil {
		return nil, err
	}
	r := reader{blocks: []domain.Block{}}
	for _, line := range pystr.SplitLinesKeepEnds(raw) {
		r.line(line)
	}
	r.flush()

	title := ""
	for _, b := range r.blocks {
		if len(b.HeadingPath) > 0 {
			title = b.HeadingPath[0]
			break
		}
	}
	if title == "" {
		title = pystr.Stem(filepath.Base(path))
	}
	return domain.NewExtraction(title, "markdown", "atx_headings", r.blocks), nil
}

// reader gathers paragraphs line by line. Offsets count code points from the
// start of the file.
type reader struct {
	stack      domain.HeadingStack
	blocks     []domain.Block
	para       []string
	paraOffset int
	offset     int
	inFence    bool
}

func (r *reader) line(line string) {
	width := utf8.RuneCountInString(line)
	defer func() { r.offset += width }()

	stripped := strings.TrimRight(line, "\n")
	if codeFence.MatchString(stripped) {
		r.inFence = !r.inFence
		r.para = append(r.para, stripped)
		return
	}
	var heading []string
	if !r.inFence {
		heading = atxHeading.FindStringSubmatch(stripped)
	}
	switch {
	case heading != nil:
		r.flush()
		r.stack.Push(len(heading[1]), pystr.Strip(heading[2]))
	case pystr.Strip(stripped) == "":
		r.flush()
		r.paraOffset = r.offset + width
	default:
		if len(r.para) == 0 {
			r.paraOffset = r.offset
		}
		r.para = append(r.para, stripped)
	}
}

func (r *reader) flush() {
	if body := pystr.Strip(strings.Join(r.para, "\n")); body != "" {
		r.blocks = append(r.blocks, domain.NewOffsetBlock(r.stack.Path(), r.paraOffset, body))
	}
	r.para = nil
}
